File: internal/remux/protocol.go
// Package remux holds the versioned contract of the remote remux pipeline:
// run states, the payloads the API, the worker and the browser exchange, and
// the codec policy. The same shapes exist in TypeScript and Rust; a change
// here is a protocol change and bumps ProtocolVersion.
package remux

import "fmt"

const ProtocolVersion = 1

const (
	RunStarting   = "starting"
	RunAccepted   = "accepted"
	RunRunning    = "running"
	RunDraining   = "draining"
	RunCompleted  = "completed"
	RunCancelling = "cancelling"
	RunCancelled  = "cancelled"
	RunFailed     = "failed"
)

func TerminalState(state string) bool {
	return state == RunCompleted || state == RunCancelled || state == RunFailed
}

// StartRequest is POST /api/torrents/:jobId/remux. Authorization is exactly
// one of OwnerToken (room bootstrap) or MemberID+Capability (connected
// controller), and both require the session that owns the torrent job.
type StartRequest struct {
	RoomID          string    `json:"roomId" binding:"required"`
	MediaGeneration int       `json:"mediaGeneration"`
	RequestID       string    `json:"requestId" binding:"required"`
	StartMs         int64     `json:"startMs"`
	Auth            StartAuth `json:"auth"`
}

// StartAuth carries exactly one of the two authorization modes.
type StartAuth struct {
	OwnerToken string `json:"ownerToken,omitempty"`
	MemberID   string `json:"memberId,omitempty"`
	Capability string `json:"capability,omitempty"`
}

// StartResponse is the 202 body: accepted is not ready, and not running.
type StartResponse struct {
	RunID           string `json:"runId"`
	MediaGeneration int    `json:"mediaGeneration"`
	State           string `json:"state"`
}

// Spec is the remux block of a signed worker job (kind "remuxStart"): only
// the payload the remux supervisor consumes, under the envelope's signature.
type Spec struct {
	ProtocolVersion int    `json:"protocolVersion"`
	RunID           string `json:"runId"`
	Claim           string `json:"claim"`
	MediaGeneration int    `json:"mediaGeneration"`
	Region          int    `json:"region"`
	StartMs         int64  `json:"startMs"`
	EndMs           int64  `json:"endMs,omitempty"`
	APIBase         string `json:"apiBase"`
	RoomID          string `json:"roomId"`
	Limits          Limits `json:"limits"`
}

// Limits are the backend-signed ceilings for one run. The worker clamps
// each to its own local configuration: a signed limit can lower a local
// ceiling, never raise it.
type Limits struct {
	PutConcurrency int   `json:"putConcurrency"`
	SpoolBytes     int64 `json:"spoolBytes"`
	ObjectBytes    int64 `json:"objectBytes"`
	AheadMs        int64 `json:"aheadMs"`
}

// Capability is what a worker announces in its heartbeat when it can run
// remote remux. Absent means it cannot, and it receives no remux jobs.
type Capability struct {
	ProtocolVersion int          `json:"protocolVersion"`
	Slots           int          `json:"slots"`
	ActiveRuns      int          `json:"activeRuns"`
	FFmpeg          string       `json:"ffmpeg"`
	AudioCodecs     []string     `json:"audioCodecs"`
	Youtube         *Youtube     `json:"youtube,omitempty"`
	Runs            []RunReport  `json:"runs,omitempty"`
	LiveSlots       int          `json:"liveSlots"`
	ActiveLives     int          `json:"activeLives"`
	Lives           []LiveReport `json:"lives,omitempty"`
}

// Youtube is the worker's word that it resolves YouTube links: yt-dlp is
// there, and whether it leaves through a proxy.
type Youtube struct {
	Version string `json:"version"`
	Proxied bool   `json:"proxied"`
}

// LiveState is where one live stands: starting, live, ended, or failed with
// the code the site can act on.
type LiveState struct {
	State  string `json:"state"`
	Code   string `json:"code,omitempty"`
	Detail string `json:"detail,omitempty"`
}

// LiveReport is one live the worker has on the relay, by the room it serves.
type LiveReport struct {
	RoomID string    `json:"roomId"`
	State  LiveState `json:"state"`
}

// TakesLive reports whether a live may be put on this worker: it resolves
// YouTube and has a live slot to spare.
func (c *Capability) TakesLive() bool {
	return c.TakesYoutube() && c.LiveSlots > c.ActiveLives
}

// TakesYoutube reports whether a YouTube run may be dispatched here.
func (c *Capability) TakesYoutube() bool {
	return c.Compatible() && c.Youtube != nil && c.Youtube.Version != ""
}

// RunReport is one run's state as the worker last told it.
type RunReport struct {
	RunID      string `json:"runId"`
	State      string `json:"state"`
	ProducedMs int64  `json:"producedMs"`
	Error      string `json:"error,omitempty"`
}

// Compatible reports whether the backend may dispatch a remux job here.
func (c *Capability) Compatible() bool {
	return c != nil && c.ProtocolVersion == ProtocolVersion && c.FFmpeg != "" && c.Slots > 0
}

// Event is a worker→backend status report on the worker link (type
// "remuxEvent"). Media publication does not ride here — it goes through the
// claim-authenticated publish endpoint; this is lifecycle only.
type Event struct {
	Type            string `json:"type"`
	WorkerID        string `json:"workerId"`
	JobID           string `json:"jobId"`
	RunID           string `json:"runId"`
	MediaGeneration int    `json:"mediaGeneration"`
	Seq             int64  `json:"seq"`
	State           string `json:"state"`
	Error           string `json:"error,omitempty"`
	ProducedMs      int64  `json:"producedMs,omitempty"`
}

type VideoVerdict int

const (
	VideoCopy VideoVerdict = iota
	VideoReject
)

// VideoPolicy copies H264, HEVC, AV1 and VP9 — whether a device decodes them
// is the player's problem — and rejects everything else: no hidden video
// transcode. Mirrored by the worker's plan.rs.
func VideoPolicy(codec string) VideoVerdict {
	switch codec {
	case "h264", "avc":
		return VideoCopy
	case "hevc", "h265":
		return VideoCopy
	case "av1", "vp9":
		return VideoCopy
	default:
		return VideoReject
	}
}

type AudioVerdict int

const (
	AudioCopy AudioVerdict = iota
	AudioConvert
	AudioReject
)

const maxAudioChannels = 8

// AudioPolicy decides copy/convert/reject for one track: AAC copies, the
// codecs with a matrix entry convert to AAC, anything else is refused.
func AudioPolicy(codec string, channels int) (AudioVerdict, error) {
	if channels <= 0 || channels > maxAudioChannels {
		return AudioReject, fmt.Errorf("audio layout with %d channels is outside the supported range", channels)
	}
	switch codec {
	case "aac":
		return AudioCopy, nil
	case "ac3", "eac3", "dts", "dca", "opus", "flac", "mp3", "vorbis":
		return AudioConvert, nil
	default:
		return AudioReject, fmt.Errorf("audio codec %q has no matrix entry", codec)
	}
}

// AACBitrateFor is the conversion bitrate per validated layout, in bits per
// second. Stereo and mono ride 160k; 5.1 and up ride 384k.
func AACBitrateFor(channels int) int {
	if channels <= 2 {
		return 160_000
	}
	return 384_000
}

File: ss-worker/ss-remux/src/protocol.rs
use serde::Deserialize;

pub const PROTOCOL_VERSION: u32 = 1;

/// The remux block of a signed `remuxStart` job. No Debug derive: `claim`
/// must never reach a log.
#[derive(Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Spec {
    pub protocol_version: u32,
    pub run_id: String,
    pub claim: String,
    #[serde(default)]
    #[allow(dead_code)]
    pub metadata_token: Option<String>,
    pub media_generation: u64,
    pub region: u64,
    #[serde(default)]
    pub start_ms: u64,
    #[serde(default)]
    pub end_ms: u64,
    pub api_base: String,
    pub room_id: String,
    #[serde(default)]
    pub limits: Limits,
}

impl std::fmt::Debug for Spec {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Spec")
            .field("run_id", &self.run_id)
            .field("media_generation", &self.media_generation)
            .field("region", &self.region)
            .field("start_ms", &self.start_ms)
            .field("room_id", &self.room_id)
            .finish_non_exhaustive()
    }
}

/// Server-signed ceilings: a signed limit can lower a local ceiling, never
/// raise it. Zero means the local value stands.
#[derive(Clone, Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Limits {
    #[serde(default)]
    pub put_concurrency: usize,
    #[serde(default)]
    pub spool_bytes: u64,
    #[serde(default)]
    pub object_bytes: u64,
    #[serde(default)]
    #[allow(dead_code)]
    pub ahead_ms: u64,
}

impl Limits {
    pub fn clamp_puts(&self, local: usize) -> usize {
        if self.put_concurrency == 0 { local } else { self.put_concurrency.min(local) }
    }
    pub fn clamp_spool(&self, local: u64) -> u64 {
        if self.spool_bytes == 0 { local } else { self.spool_bytes.min(local) }
    }
    pub fn clamp_object(&self, local: u64) -> u64 {
        if self.object_bytes == 0 { local } else { self.object_bytes.min(local) }
    }
}

/// Run states, mirrored from internal/remux.
pub mod state {
    pub const ACCEPTED: &str = "accepted";
    pub const RUNNING: &str = "running";
    pub const DRAINING: &str = "draining";
    pub const COMPLETED: &str = "completed";
    pub const CANCELLED: &str = "cancelled";
    pub const FAILED: &str = "failed";
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn limits_lower_never_raise() {
        let l = Limits { put_concurrency: 8, spool_bytes: 0, object_bytes: 1024, ahead_ms: 0 };
        assert_eq!(l.clamp_puts(4), 4);
        assert_eq!(l.clamp_puts(16), 8);
        assert_eq!(l.clamp_spool(999), 999);
        assert_eq!(l.clamp_object(4096), 1024);
    }

    #[test]
    fn debug_never_prints_the_claim() {
        let spec: Spec = serde_json::from_value(serde_json::json!({
            "protocolVersion": 1, "runId": "r", "claim": "SECRET",
            "mediaGeneration": 0, "region": 0, "apiBase": "http://a", "roomId": "x"
        }))
        .unwrap();
        assert!(!format!("{spec:?}").contains("SECRET"));
    }
}
